### ISSUE TITLE
Iterating over chunks efficiently

### ISSUE THREAD

**Issue by k-cybulski** [JuliaData/CSV.jl#482]
I'm trying to parse a file too large to fit in memory by chunks and I'm having issues with doing so efficiently. For reference, in pandas `read_csv` has a `chunksize` argument that does roughly what I'm trying to do [(see here)](https://pandas.pydata.org/pandas-docs/stable/reference/api/pandas.read_csv.html#pandas-read-csv)

My current approach is to iterate by `skipto` and `limit` like so:
```julia
data_iterator(chunksize::Int64, chunks::Int64) = (
    CSV.read(DATA_PATH, skipto=firstrow, limit=chunksize, header=DATA_COLUMNS) 
    for firstrow in 1:chunksize:chunksize*chunks)
```
and
```julia
for df in data_iterator(1000, 10)
    doThings(df)
end
```

The issue is `skipto` slows down considerably the deeper my iterator gets in the file, as seen in this micro benchmark:
```
julia> mean([@elapsed CSV.File(filePath, limit=1000) for i in 1:5000])
0.0014377240542

julia> mean([@elapsed CSV.File(filePath, skipto=1000, limit=1000) for i in 1:5000])
0.0017316926776

julia> mean([@elapsed CSV.File(filePath, skipto=5000, limit=1000) for i in 1:5000])
0.0030615413232

julia> mean([@elapsed CSV.File(filePath, skipto=10000, limit=1000) for i in 1:5000])
0.004779332284999999
```

Is there a better way to iterate over chunks, such that I don't increase
overhead by skipping the same lines over and over again?

**Comment by quinnj** [comment]
Please see `CSV.Rows`, which is built to be a low-overhead row iterator

**Comment by k-cybulski** [comment]
Thanks, `CSV.Rows` looks like exactly what I need for some more complicated data iterators in my project.

Nonetheless, I think a pandas-esque `chunksize` chunking interface would be very handy for `CSV.jl`.

**Comment by xiaodaigh** [comment]
> I think a pandas-esque `chunksize` chunking interface would be very handy for `CSV.jl`.

Would it be easy to build a chunksize iterator on top of CSV.Rows?

**Comment by quinnj** [comment]
@k-cybulski, I'm not familiar w/ `chunksize` in pandas; what does it do? Just wondering what is easier/better about iterating chunks instead of just iterating rows.

**Comment by xiaodaigh** [comment]
sometimes we want to collect like 1millions lines and append it to a database. It usually to do it in chunks to enable more efficient buffering etc. Is loading rows efficient in CSV.jl? Does it load a large amounts of data into a buffer, but only serves the row by row via an interface? Or does it not need that to be efficient?

**Comment by quinnj** [comment]
Using `CSV.Rows` is very efficient; it mmaps the entire file initially, but then uses a row buffer when iterating. If you pass `reusebuffer=true`, it will reuse a single row buffer on each iteration. It does the cell parsing on each iteration, so yes, it's extremely efficient and only does the work as the `CSV.Rows` object is iterated.

**Comment by xiaodaigh** [comment]
> If you pass `reusebuffer=true`, it will reuse a single row buffer on each iteration.

Should this be the default? It sounds more memory efficient.

**Comment by quinnj** [comment]
It was considered, but if someone does `rows = collect(CSV.Rows(file))`, then `reusebuffer=true` results in an entire array of the last row iterated (since the same row buffer is used by each `CSV.Row` object). So for safety and avoiding confusion, the default is `reusebuffer=false`, but if you're just iterating `CSV.Rows` and don't need to look back or ahead, then `reusebuffer=true` is an easy way to increase efficiency.

**Comment by k-cybulski** [comment]
@quinnj In pandas, when you use the keyword argument `chunksize` of `read_csv`, what you get instead of a dataframe is an iterator object that reads and yields smaller dataframes of given size. [Here](https://pandas.pydata.org/pandas-docs/stable/user_guide/io.html#iterating-through-files-chunk-by-chunk) is a little demonstration.

In my specific use case, the disjoint segments of a massive CSV I want to analyse are of variable size, so an iterator that allows for multiple chunk lengths would be ideal. As in, `for df in csvIterator(filePath, chunks=[chunk_2_start, chunk_3_start])` which would yield chunks from line 1 to `chunk_2_start - 1`, from `chunk_2_start` to `chunk_3_start - 1` and from `chunk_3_start` to the end. This is probably a tad too niche, though.

**Comment by xiaodaigh** [comment]
> pandas

Problem with pandas is that it's dead slow. I think there is a chunking interface being worked on. In practice, I think the best way to deal with large CSV is to split them up using `split` and then read them in in parallel.

**Comment by quinnj** [comment]
Ok, I've thought about this issue a few times for quite a while now and IMO, the best way to work through chunks of a file is using a combination of existing tools:

* `CSV.Rows` is an efficient iterator over a csv file, only buffering a single row at a time
* `Iterators.partition(collection, n)` allows "chunking" an iterator in chunks of `n`
* `DataFrame(rows)` allows converting a chunk of csv rows into a DataFrame if that is more convenient for working with

The only other piece of functionality we might consider is allowing "chunking" a `CSV.Rows` object into `n` chunks, equally spaced throughout the file; this could allow then having multiple threads handling each chunk. But that's also not a paradigm I've heard of much, vs. just doing `Threads.@spawn` on each row while iterating to process, or doing the `Iterators.partition` with `Threads.@spawn` to have threads process chunks. At the moment, I don't think CSV.jl providing something will stand the test of time, so I'd rather not complicate things when existing paradigms/functionality should support various workflows.

**Comment by xiaodaigh** [comment]
In case someone googles lands here. Here is what @quinnj is suggest you do

```julia
using CSV

using Base.Iterators: partition
using DataFrames: DataFrame

using RDatasets

mtcars = dataset("datasets", "mtcars")
CSV.write("mtcars.csv", mtcars)

read_chunks(filepath, n = 1_000_000; csv_rows_params...) = begin
    cr = CSV.Rows(filepath; csv_rows_params...)
    # put in some custom code to apply the right type here.
    [DataFrame(chunk) for chunk in Iterators.partition(cr, n)]
end

filepath = "mtcars.csv"

dfs = read_chunks(filepath, 5)

# no promote type; makes no difference cos everything  is `String`
df1 = reduce(append!, dfs)

# promote type; only helps after type inference code added.
df2 = reduce(vcat, dfs)
```

Now, `dfs` is a list of DataFrame potentially where each column is of a different type and `df`.

As of 2020-04-17, here are some timings `https://docs.rapids.ai/datasets/mortgage-data`.

In particular the `Performance_2000Q4.txt` dataset

| Method | Speed | 
| -- | -- |
| `CSV.Rows` no type inference |  53.74 |
| `CSV.read` | 8.55 |
| `CSV.Rows` with type inference |  ??? |

Presumably you can run an algorithm like using `CSV.read` on a subset of rows to infer type, and then apply the types.

**Comment by quinnj** [comment]
Hmmmm, this isn't quite what I had in mind:
* Just doing `[DataFrame(chunk) for chunk in Iterators.partition(cr, n)]` seems.....wasteful to me? Like, if you need to chunk a file, then this is just materializing the full table/file anyway, just in `n` DataFrames instead of one? I was imagining more a scenario where you do:

```julia
for rows in Iterators.partition(CSV.Rows(file), n)
    # df = DataFrame(rows); # materialize a DataFrame if you need it
    # perform operations on this specific chunk of file
end
```
Do note that calling `DataFrame(rows)` isn't free and will "build columns" from the rows; I don't think it's prohibitive or should be discouraged, because sometimes a DataFrame is the best tool for the job, but just note that if you can do your operations/processing on the rows themselves, that will be cheaper.

* I'm note sure what you mean by the comment `# put in some custom code to apply the right type here.` There is code on master that allows you to pass `types` to `CSV.Rows` in order to have columns be natively parsed to a specified type; you can also do `CSV.detect(row, i)` or `Parsers.parse(row, T, i)` to infer the type of a cell in a row (former) or parse a specific type (latter). But there's nothing you can do to a `CSV.Rows` to "apply the right type".

* Again, I'm not sure why you're suggesting to do `df1 = reduce(append!, dfs)`? This just materializes one huge DataFrame from the individual DataFrame chunks? If you really want the full DataFrame, just do `DataFrame(CSV.File(file))`, right?

`CSV.Rows` exists purely for iteration, low-memory use-cases, and, as I suggested above, can be used in conjunction with `Iterators.partition` to chunk up, but I would discourage any use where you end up materializing all rows in any way.

**Comment by xiaodaigh** [comment]
> * ust doing `[DataFrame(chunk) for chunk in Iterators.partition(cr, n)]`

That's just an illustration.

**Comment by xiaodaigh** [comment]
The key for me is iterating, and I want each chunk to be returned as dataframe. I think building an iterator where 

```
for chunk in CsvChunkIterator(file)
  # chunk is of type DataFrame with column types inferred
end
```

**Comment by xiaodaigh** [comment]
Ok Got it. This is going into https://github.com/xiaodaigh/DataConvenience.jl


```julia
struct CsvChunksIterator
    chunk_iterator::Base.Iterators.PartitionIterator{T} where T

    CsvChunksIterator(path::String, n = 2^16) = begin
        # firstly read some columns to infer types
        df = CSV.read(path, limit = n)
        types = [eltype(col[1:1]) for col in eachcol(df)]
        csv_rows = CSV.Rows(path, types = types)
        new(Iterators.partition(csv_rows, n))
    end
end
```

**Comment by xiaodaigh** [comment]
Trying to read a 800mb file, the `CSV.read` performance is really good at 8s-11s. However, if I try to read it by chunk using a combination of `CSV.Rows` and `Iterators.partition`, it takes 2mins. 

It feels like `CSV.Rows` is meant for fairly low level use and `CSV.read` is high level. I am trying to get something in between. I am trying to make an Iterator so that each iteration returns a `DataFrame` with types `inferred`. If `CSV.read` can return a dataframe with types inferred, then I think the chunk version should not be 12x slower. Any hint on how to achieve it? `DataFrame`ing the `CSV.Rows` from a partition isn't the way. See cod below 

```julia
using CSV

using Base.Iterators: partition
using DataFrames: DataFrame

filepath = "c:/data/perf/Performance_2004Q3.txt"

using BenchmarkTools
@benchmark CSV.read(filepath) #8s

time_chunks(filepath, n = 100_000; csv_rows_params...) = begin
    cr = CSV.Rows(filepath; csv_rows_params...)
    # put in some custom code to apply the right type here.
    for chunk in Iterators.partition(cr, n)
        DataFrame(chunk)
    end
end

@benchmark time_chunks(filepath) #2 mins
```

**Comment by quinnj** [comment]
One thing we could expose is that if you do `f = CSV.File(file)` w/ multiple threads, we could have a function like `CSV.chunks(f)` that would take a full `CSV.File` and split it into N `CSV.File`s, one for each thread that was used for parsing. That would allow processing the "chunks", and converting each to a DataFrame efficiently.

**Comment by xiaodaigh** [comment]
That would be pretty awesome!

**Comment by quinnj** [comment]
Ok, wanna give https://github.com/JuliaData/CSV.jl/pull/615 a try?
